src/main/java/io/cdap/plugin/MultiTableDBSource.java

```diff
@@ -73,19 +87,72 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
   public void prepareRun(BatchSourceContext context) throws Exception {
     Configuration hConf = new Configuration();
     Class<? extends Driver> driverClass = context.loadPluginClass(JDBC_PLUGIN_ID);
-    Collection<DBTableInfo> tables = MultiTableDBInputFormat.setInput(hConf, conf, driverClass);
-    SettableArguments arguments = context.getArguments();
-    for (DBTableInfo tableInfo : tables) {
-      arguments.set(DynamicMultiFilesetSink.TABLE_PREFIX + tableInfo.getDbTableName().getTable(),
-                    tableInfo.getSchema().toString());
-    }
 
-    context.setInput(Input.of(conf.getReferenceName(),
-                              new SourceInputFormatProvider(MultiTableDBInputFormat.class, hConf)));
+    Collection<DBTableInfo> tables;
+
+    try {
+      tables = MultiTableDBInputFormat.setInput(hConf, conf, driverClass);
+
+      SettableArguments arguments = context.getArguments();
+      for (DBTableInfo tableInfo : tables) {
+        arguments.set(DynamicMultiFilesetSink.TABLE_PREFIX + tableInfo.getDbTableName().getTable(),
+                      tableInfo.getSchema().toString());
+      }
+
+      context.setInput(Input.of(conf.getReferenceName(),
+                                new SourceInputFormatProvider(ErrorCollectingMultiTableDBInputFormat.class, hConf)));
+    } catch (Exception ex) {
+      String errorMessage = "Error getting table schemas from database.";
+      LOG.error(errorMessage, ex);
+
+      InputFormatProvider errorEmittingInputFormatProvider = new InputFormatProvider() {
+        @Override
+        public String getInputFormatClassName() {
+          return ErrorEmittingInputFormat.class.getName();
+        }
+
+        @Override
+        public Map<String, String> getInputFormatConfiguration() {
+          Map<String, String> config = new HashMap<>();
+          for (Map.Entry<String, String> entry : hConf) {
+            config.put(entry.getKey(), entry.getValue());
+          }
+          //Add ErrorEmittingInputFormat config properties
+          config.put(ErrorEmittingInputFormat.ERROR_MESSAGE, errorMessage);
+          config.put(ErrorEmittingInputFormat.EXCEPTION_CLASS_NAME, ex.getClass().getCanonicalName());
+
+          return config;
+        }
+      };
+
+      context.setInput(Input.of(conf.getReferenceName(), errorEmittingInputFormatProvider));
+    }
   }
 
   @Override
   public void transform(KeyValue<NullWritable, StructuredRecord> input, Emitter<StructuredRecord> emitter) {
-    emitter.emit(input.getValue());
+    StructuredRecord record = input.getValue();
+
+    // Check if the record is an error.
+    if (ErrorSchema.SCHEMA_NAME.equals(record.getSchema().getRecordName())) {
```

Comment:
instead of checking the schema name, I think it would be cleaner if the InputFormat instead returns a 'RecordWrapper' (can think of a better name) object that has 2 variables, one InvalidEntry and one StructuredRecord. That way the RecordReader could potentially use different error codes depending on what type of error it got, and we're not relying on checking a special string value.

Alternatively, could make the key an InvalidEntry instead of a NullWritable.

Reply:
I like the RecordWrapper idea.

Reply:
I have implemented the your RecordWrapper suggestion.